Principles/6 - Declarative over imperative.md

```diff
@@ -0,0 +1,27 @@
+# Declarative over imperative
+
+Imperative code is often a natural choice for developers who have spent a lot of time in C-style languages. The primitives of many popular programming languages are very easy to get started with writing highly imperative code. However, building imperative APIs can lead to much more error prone systems, more difficult to read code, and it being harder to discern the intention from the implementation for a given piece of logic.
+
+We prefer to write code that exposes a declarative API than an imperative API. A declarative API will let you define what results you would like to see, rather than the way you want them to happen. This type of API leads to simpler code with a higher degree of readability, maintainability, and composability. Where imperative code must exist we often counsel using more declarative and immutable styles within it.
+
+Below are a few examples of decisions that demonstrate this principle, some at the highest level of technology choice, and some at the smallest level of individual function implementations:
```

Comment:
other choices I thought about mentioning but didn't because they had too much overlap or seemed not super relevant

- imports over require. Import is declarative, require is imperative
-  typescript instead of runtime checks. Typescript is declarative, runtime checks are imperative
- SCSS is a declarative choice vs CSS in JS or some other crazy thing, seemed too trivial but it *is* a declarative choice
- how much we hate graphql cache manipulation - we don't have a solution yet but it's one of the worst parts of our codebase. Once we have a solution (which will assumedly be declarative) it should go here.

Reply:
I would just ditch the entire part of this paragraph starting with “, some”